### PR TITLE
Ea 2401 modellingof entitysets

### PR DESCRIPTION
**Pull request by SrishtiSingh-eu**

### CODE REVIEW

File: set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java [hunk 1]
```diff
@@ -371,4 +215,34 @@ public boolean isPublic() {
...
 	return VisibilityTypes.PUBLISHED.getJsonValue().equals(getVisibility());
     }
+
+    /**
+     * if List<String>subject contains an entity reference
+     * referring to the set then it is a EntityBestItemSet
+     *
+     * @return true if it is an EntityBestItemSet
+     */
+    public boolean isEntityBestItemSet() {
```
Comment (gsergiu) [inline set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: this method should simply verify the type of the set to be EntityBestItemSet

File: set-mongo/src/main/java/eu/europeana/set/web/model/WebUserSetImpl.java [hunk 2]
```diff
@@ -22,7 +22,7 @@
 
 @JsonPropertyOrder({ WebUserSetModelFields.ID, WebUserSetModelFields.TYPE, WebUserSetModelFields.TITLE, WebUserSetFields.DESCRIPTION,
         WebUserSetModelFields.VISIBILITY, WebUserSetModelFields.IS_DEFINED_BY, WebUserSetModelFields.ITEMS, WebUserSetModelFields.CREATOR,
-        WebUserSetModelFields.CREATED, WebUserSetModelFields.MODIFIED, WebUserSetModelFields.TOTAL, WebUserSetFields.NEXT,
+        WebUserSetModelFields.CREATED, WebUserSetModelFields.MODIFIED,WebUserSetModelFields.SUBJECT, WebUserSetModelFields.TOTAL, WebUserSetFields.NEXT,
```
Comment (gsergiu) [inline set-mongo/src/main/java/eu/europeana/set/web/model/WebUserSetImpl.java:25]: We have to ask Hugo for the ordering, but I assume that the next field should be placed before or after description

File: set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java (hunk 1 above)
Comment (gsergiu) [reply set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: there is a method isEntityBestItemSet in the service with the correct implementation. I think we can remove this one.

Comment (gsergiu) [inline set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: there is a method isEntityBestItemSet in the service with the correct implementation. I think we can remove this one.

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java [hunk 3]
```diff
@@ -488,6 +492,14 @@ void validateControlledValues(UserSet webUserSet) throws RequestBodyValidationEx
 	    throw new RequestBodyValidationException(UserSetI18nConstants.USERSET_VALIDATION_PROPERTY_VALUE,
 		    new String[] { WebUserSetModelFields.TYPE, webUserSet.getType() });
 	}
+	// if type is BookmarkFolder or Collection, validate subject
+	// subject must not contain entity reference
+	if(webUserSet.getType() != null && ! isEntityBestItemsSet(webUserSet)) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:497]: the type of userSet cannot be null here, isEntityBestItemsSet(webUserSet) is sufficient

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java [hunk 4]
```diff
@@ -488,6 +492,14 @@ void validateControlledValues(UserSet webUserSet) throws RequestBodyValidationEx
...
+		if(webUserSet.isEntityBestItemSet()) {
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:498]: the method in the userset should be removed

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java [hunk 5]
```diff
@@ -488,6 +492,14 @@ void validateControlledValues(UserSet webUserSet) throws RequestBodyValidationEx
...
+			throw new RequestBodyValidationException(UserSetI18nConstants.USERSET_VALIDATION_PROPERTY_VALUE,
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:499]: there is a different message which is specific for missing mandatory fields. Please switch to the other message by using appropriate parameters

Comment (gsergiu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:499]: Isn't this validation part of the validateBestItemsSet?

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java [hunk 6]
```diff
@@ -534,6 +546,41 @@ void validateIsDefinedBy(UserSet webUserSet) throws ParamValidationException, Re
...
+		}
+
+		if (webUserSet.isOpenSet()) {
+			throw new ParamValidationException(UserSetI18nConstants.USERSET_VALIDATION_PROPERTY_NOT_ALLOWED,
+					UserSetI18nConstants.USERSET_VALIDATION_PROPERTY_NOT_ALLOWED,
+					new String[] { WebUserSetModelFields.IS_DEFINED_BY, webUserSet.getType() });
+		}
+
+		if (!webUserSet.isEntityBestItemSet()) {
+			throw new RequestBodyValidationException(UserSetI18nConstants.USERSET_VALIDATION_ENTITY_REFERENCE,
```
Comment (gsergiu) [inline set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:575]: the validation rules (business logic) must be implemented here not in the model classes

**Review by gsergiu — CHANGES_REQUESTED** [review]
please implement the required changes

File: set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java (hunk 1 above)
Comment (SrishtiSingh-eu) [reply set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: Bt this method is to verify if the subject filed is present and has entity reference. Ther service implementation one just checks the type of the set. Either i can add this logic there or we keep this method. This something like isOpenSet

Comment (SrishtiSingh-eu) [reply set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: have moved the business logic to service class

File: set-mongo/src/main/java/eu/europeana/set/web/model/WebUserSetImpl.java (hunk 2 above)
Comment (SrishtiSingh-eu) [reply set-mongo/src/main/java/eu/europeana/set/web/model/WebUserSetImpl.java:25]: have placed before description

File: set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java (hunk 1 above)
Comment (SrishtiSingh-eu) [reply set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java:225]: done

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java (hunk 3 above)
Comment (SrishtiSingh-eu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:497]: done

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java (hunk 4 above)
Comment (SrishtiSingh-eu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:498]: done

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java (hunk 5 above)
Comment (SrishtiSingh-eu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:499]: this validation is for BookmarkFolder and Collection type user set. As they cannot have a entity reference in subject field

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java (hunk 6 above)
Comment (SrishtiSingh-eu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java:575]: done

File: set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java (hunk 5 above)
Comment (SrishtiSingh-eu) [reply set-web/src/main/java/eu/europeana/set/web/service/impl/BaseUserSetServiceImpl.java]: Have changed the message

File: set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java (hunk 1 above)
Comment (gsergiu) [reply set-definitions/src/main/java/eu/europeana/set/definitions/model/impl/BaseUserSet.java]: the isOpenSet is somethinf different, becasue it has the same type as the closed set. But entity sets have their own type

**Pull request merged**